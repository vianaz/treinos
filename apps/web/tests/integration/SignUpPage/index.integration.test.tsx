import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ThemeProvider } from 'styled-components';

import { SignUp } from '@templates';
import { themeLight } from '@themes';
import { faker } from '@faker-js/faker';

const makeSut = () => {
  return render(
    <ThemeProvider theme={themeLight}>
      <SignUp />
    </ThemeProvider>
  );
};
describe('Sign Up Page', () => {
  vi.mock('next/router', () => ({
    useRouter() {
      return {
        push: vi.fn()
      };
    }
  }));

  describe('Render Tests', () => {
    it('should render sign up page correct', () => {
      const { container, getByText } = makeSut();
      const texts = [
        'email',
        'senha',
        'data de nascimento',
        'Cadastrar'
      ];

      expect(container).toBeInTheDocument();
      texts.forEach(text => {
        expect(getByText(text)).toBeInTheDocument();
      });
      expect(container).toMatchSnapshot();
    });
  });

  describe('Email Input Tests', () => {
    it('should write correct in email input', () => {
      const { getAllByTestId } = makeSut();
      const email = faker.internet.email();

      const inputs = getAllByTestId('input');

      const emailInput = inputs[0];

      act(() => userEvent.type(emailInput, email));

      expect(emailInput).toHaveValue(email);
    });

    it('should show error message if email is empty', () => {
      const { getAllByTestId, getByText, getByTestId } = makeSut();

      const inputs = getAllByTestId('input');
      const button = getByTestId('button');

      const emailInput = inputs[0];

      act(() => userEvent.click(button));

      expect(getByText('*Email é obrigatório')).toBeInTheDocument();
      expect(emailInput).toHaveValue('');
    });

    it('should show error message if email is invalid', () => {
      const { getAllByTestId, getByText, getByTestId } = makeSut();
      const email = faker.random.word();

      const inputs = getAllByTestId('input');
      const button = getByTestId('button');
      const emailInput = inputs[0];

      act(() => {
        userEvent.type(emailInput, email);
        userEvent.click(button);
      });

      expect(emailInput).toHaveValue(email);
      expect(getByText('*Email inválido')).toBeInTheDocument();
      expect(getByText('*Email inválido')).toHaveStyle({
        color: themeLight.colors.primary[900]
      });
    });
  });

  describe('Password Input Tests', () => {
    it('should write correct in password input', () => {
      const { getAllByTestId, getByTestId, queryByText } = makeSut();
      const password = faker.internet.password(8);

      const inputs = getAllByTestId('input');
      const emailInput = inputs[0];
      const passwordInput = inputs[1];
      const button = getByTestId('button');

      act(() => {
        userEvent.type(emailInput, faker.internet.email());
        userEvent.type(passwordInput, password);
        userEvent.click(button);
      });

      expect(passwordInput).toHaveValue(password);
      expect(
        queryByText('*Senha é obrigatória')
      ).not.toBeInTheDocument();
    });

    it('should show error message if password is empty', () => {
      const { getAllByTestId, getByText, getByTestId } = makeSut();

      const inputs = getAllByTestId('input');
      const emailInput = inputs[0];
      const passwordInput = inputs[1];
      const button = getByTestId('button');

      act(() => {
        userEvent.type(emailInput, faker.internet.email());
        userEvent.click(button);
      });

      expect(getByText('*Senha é obrigatória')).toBeInTheDocument();
      expect(passwordInput).toHaveValue('');
    });

    it('should show error message if password is invalid', () => {
      const { getAllByTestId, getByText, getByTestId } = makeSut();
      const password = faker.internet.password(7);

      const inputs = getAllByTestId('input');
      const emailInput = inputs[0];
      const passwordInput = inputs[1];
      const button = getByTestId('button');

      act(() => {
        userEvent.type(emailInput, faker.internet.email());
        userEvent.type(passwordInput, password);
        userEvent.click(button);
      });

      expect(passwordInput).toHaveValue(password);
      expect(
        getByText('*A senha deve ter no mínimo 8 caracteres')
      ).toBeInTheDocument();
      expect(
        getByText('*A senha deve ter no mínimo 8 caracteres')
      ).toHaveStyle({
        color: themeLight.colors.primary[900]
      });
    });
  });

  describe('Born Date Input Tests', () => {
    it('should write correct in born date input', () => {
      const { getAllByTestId, getByTestId, queryByText } = makeSut();
      const bornDate = faker.date.past().toISOString().split('T')[0];

      const inputs = getAllByTestId('input');
      const emailInput = inputs[0];
      const passwordInput = inputs[1];
      const bornDateInput = inputs[2];
      const button = getByTestId('button');

      act(() => {
        userEvent.type(emailInput, faker.internet.email());
        userEvent.type(passwordInput, faker.internet.password(8));
        userEvent.type(bornDateInput, bornDate);
        userEvent.click(button);
      });

      expect(bornDateInput).toHaveValue(bornDate);
      expect(
        queryByText('*Data de nascimento é obrigatória')
      ).not.toBeInTheDocument();
    });

    it('should show error message if born date is empty', () => {
      const { getAllByTestId, getByText, getByTestId } = makeSut();

      const inputs = getAllByTestId('input');
      const emailInput = inputs[0];
      const passwordInput = inputs[1];
      const button = getByTestId('button');

      act(() => {
        userEvent.type(emailInput, faker.internet.email());
        userEvent.type(passwordInput, faker.internet.password(8));
        userEvent.click(button);
      });

      expect(
        getByText('*Data de nascimento é obrigatória')
      ).toBeInTheDocument();
    });

    it('should show error message if born date is invalid', () => {
      const { getAllByTestId, getByText, getByTestId } = makeSut();
      const bornDate = faker.date
        .future()
        .toISOString()
        .split('T')[0];

      const inputs = getAllByTestId('input');
      const emailInput = inputs[0];
      const passwordInput = inputs[1];
      const bornDateInput = inputs[2];
      const button = getByTestId('button');

      act(() => {
        userEvent.type(emailInput, faker.internet.email());
        userEvent.type(passwordInput, faker.internet.password(8));
        userEvent.type(bornDateInput, bornDate);
        userEvent.click(button);
      });

      expect(bornDateInput).toHaveValue(bornDate);
      expect(
        getByText('*Você deve ter no mínimo 18 anos')
      ).toBeInTheDocument();
      expect(
        getByText('*Você deve ter no mínimo 18 anos')
      ).toHaveStyle({
        color: themeLight.colors.primary[900]
      });
    });
  });
});
