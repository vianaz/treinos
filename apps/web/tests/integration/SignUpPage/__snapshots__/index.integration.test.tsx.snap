// Vitest Snapshot v1

exports[`Sign Up Page > Render Tests > should render sign up page correct 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 16rem;
  padding: 0.5rem;
  margin: 0px 10px;
  border: 1px solid #BBBBBB;
  border-radius: 0.5rem;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 1rem;
  -webkit-letter-spacing: 0.07rem;
  -moz-letter-spacing: 0.07rem;
  -ms-letter-spacing: 0.07rem;
  letter-spacing: 0.07rem;
  border: none;
  background: transparent;
}

.c5:focus {
  outline: none;
}

.c3 {
  -webkit-letter-spacing: 0.05rem;
  -moz-letter-spacing: 0.05rem;
  -ms-letter-spacing: 0.05rem;
  letter-spacing: 0.05rem;
  font-size: 1rem;
  color: #BBBBBB;
  font-weight: 400;
  font-family: 'Open Sans',sans-serif;
  margin-bottom: 0.5rem;
  padding-left: 1rem;
}

.c6 {
  width: 190px;
  height: 64px;
  aspect-ratio: 21 / 9;
  border-radius: 10px;
  border: none;
  background-color: #FF4800;
}

.c6:hover {
  cursor: pointer;
}

.c7 {
  font-size: 1.25rem;
  font-family: 'Open Sans',sans-serif;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  color: #fff;
}

.c8 {
  font-size: 0.938rem;
  font-family: 'Open Sans',sans-serif;
  -webkit-letter-spacing: 0.1rem;
  -moz-letter-spacing: 0.1rem;
  -ms-letter-spacing: 0.1rem;
  letter-spacing: 0.1rem;
  color: #BBBBBB;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 18rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 2.5rem;
}

.c1 a {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #FF5A1A;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 100vh;
}

<div>
  <main
    class="c0"
  >
    <form
      class="c1"
    >
      <div
        class="c2"
      >
        <label
          class="c3"
          data-testid="label-input"
          for="text"
        >
          email
        </label>
        <div
          class="c4"
        >
          <input
            class="c5"
            data-testid="input"
            name="email"
            placeholder="email@treinÃ³s.com"
            type="text"
          />
        </div>
      </div>
      <div
        class="c2"
      >
        <label
          class="c3"
          data-testid="label-input"
          for="password"
        >
          senha
        </label>
        <div
          class="c4"
        >
          <input
            class="c5"
            data-testid="input"
            name="password"
            placeholder="********"
            type="password"
          />
          <img
            alt="show password"
            data-nimg="1"
            decoding="async"
            height="20"
            loading="lazy"
            src="/icons/eye_light.svg"
            srcset="/icons/eye_light.svg 1x, /icons/eye_light.svg 2x"
            style="color: transparent;"
            width="20"
          />
        </div>
      </div>
      <div
        class="c2"
      >
        <label
          class="c3"
          data-testid="label-input"
          for="date"
        >
          data de nascimento
        </label>
        <div
          class="c4"
        >
          <input
            class="c5"
            data-testid="input"
            name="bornDate"
            placeholder="dd/mm/aaaa"
            type="date"
          />
        </div>
      </div>
      <button
        class="c6"
        data-testid="button"
      >
        <p
          class="c7"
          type="primary"
        >
          Cadastrar
        </p>
      </button>
      <p
        class="c8"
        type="tertiary"
      >
        Ja tem uma conta? Faca 
        <a
          href="/signin"
        >
          login
        </a>
      </p>
    </form>
  </main>
</div>
`;
